azure/services/securitygroups/securitygroups.go

```diff
@@ -52,111 +54,72 @@ func New(scope NSGScope) *Service {
 	}
 }
 
-// Reconcile gets/creates/updates a network security group.
+// Reconcile gets/creates/updates network security groups.
 func (s *Service) Reconcile(ctx context.Context) error {
 	ctx, span := tele.Tracer().Start(ctx, "securitygroups.Service.Reconcile")
 	defer span.End()
 
-	if !s.Scope.IsVnetManaged() {
-		s.Scope.V(4).Info("Skipping network security group reconcile in custom VNet mode")
+	ctx, cancel := context.WithTimeout(ctx, reconciler.DefaultAzureServiceReconcileTimeout)
+	defer cancel()
+
+	// Only create the NSGs if their lifecycle is managed by this controller.
+	// NSGs are managed if and only if the vnet is managed.
+	managed, err := s.Scope.IsVnetManaged()
+	if err != nil {
+		return errors.Wrap(err, "failed to determine if network security groups are managed")
+	} else if !managed {
+		s.Scope.V(4).Info("Skipping network security groups reconcile in custom VNet mode")
 		return nil
 	}
 
+	// We go through the list of NSGSpecs to reconcile each one, independently of the result of the previous one.
+	// If multiple erros occur, we return the most pressing one
+	// order of precedence is: error creating -> creating in progress -> created (no error)
+	var result error
+
 	for _, nsgSpec := range s.Scope.NSGSpecs() {
-		securityRules := make([]network.SecurityRule, 0)
-		var etag *string
-
-		existingNSG, err := s.client.Get(ctx, s.Scope.ResourceGroup(), nsgSpec.Name)
-		switch {
-		case err != nil && !azure.ResourceNotFound(err):
-			return errors.Wrapf(err, "failed to get NSG %s in %s", nsgSpec.Name, s.Scope.ResourceGroup())
-		case err == nil:
-			// security group already exists
-			// We append the existing NSG etag to the header to ensure we only apply the updates if the NSG has not been modified.
-			etag = existingNSG.Etag
-			// Check if the expected rules are present
-			update := false
-			securityRules = *existingNSG.SecurityRules
-			for _, rule := range nsgSpec.SecurityRules {
-				sdkRule := converters.SecurityRuleToSDK(rule)
-				if !ruleExists(securityRules, sdkRule) {
-					update = true
-					securityRules = append(securityRules, sdkRule)
-				}
-			}
-			if !update {
-				// Skip update for NSG as the required default rules are present
-				s.Scope.V(2).Info("security group exists and no default rules are missing, skipping update", "security group", nsgSpec.Name)
-				continue
-			}
-		default:
-			s.Scope.V(2).Info("creating security group", "security group", nsgSpec.Name)
-			for _, rule := range nsgSpec.SecurityRules {
-				securityRules = append(securityRules, converters.SecurityRuleToSDK(rule))
+		if err := async.CreateResource(ctx, s.Scope, s.client, nsgSpec, serviceName); err != nil {
+			if !azure.IsOperationNotDoneError(err) || result == nil {
+				result = err
```

Comment:
Maybe we could add some tests to capture the multiple errors scenario? As it's a bit of a tricky logic in here. Multiple errors can be a mix of operation-not-done errors and non operation-not-done errors and in any order. Depending on the order of errors and the kind of error, the result gets overwritten multiple times or it doesn't get overwritten